File: Difficulty: Medium/Longest Subarray with Majority Greater than K/longest-subarray-with-majority-greater-than-k.cpp
class Solution {
  public:
    int longestSubarray(vector<int> &arr, int k) {
        int n=arr.size();
        int sum=0;
        int ans=0;
        unordered_map<int,int>mp;
        mp[0]=-1;
        for(int i=0;i<n;i++){
            sum+=(arr[i]>k?1:-1);
            if(sum>0){
                ans=i+1;
            }
            
            if(mp.find(sum-1)!=mp.end()){
                ans=max(ans,i-mp[sum-1]);
            }
            if(mp.find(sum)==mp.end()){
                mp[sum]=i;
            }
        }
        return ans;
    }
};

File: Difficulty: Medium/Longest Subarray with Majority Greater than K/longest-subarray-with-majority-greater-than-k.java
class Solution {
    public int longestSubarray(int[] arr, int k) {
        HashMap<Integer, Integer> map = new HashMap<>();

        int maxLen = 0;
        int n = arr.length;

        int gtk = 0;  // count of elements greater than k
        int lek = 0;  // count of elements less than or equal to k

        int j = 0;

        while (j < n) {

            // Update counts
            if (arr[j] > k) {
                gtk++;
            } else {
                lek++;
            }

            int diff = gtk - lek;

            /*
             We check (diff - 1) because we assume that
             one more <= k element may have appeared in a previous window.
            */

            if ((diff - 1) < 0 && map.containsKey(diff - 1)) {
                maxLen = Math.max(maxLen, j - map.get(diff - 1));
            } else {
                if ((diff - 1) >= 0) {
                    maxLen = Math.max(maxLen, j + 1);
                }
            }

            // Store first occurrence of diff
            if (!map.containsKey(diff)) {
                map.put(diff, j);
            }

            j++;
        }

        return maxLen;
    }
}